zkevm-circuits/src/exp_circuit.rs

```diff
@@ -382,37 +401,137 @@ impl<F: Field> ExpCircuitConfig<F> {
 
         Ok(())
     }
+    /// Assign witness to the exponentiation circuit.
+    pub fn assign_exp_events(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        exp_events: &[ExpEvent],
+        max_exp_steps: usize,
+    ) -> Result<(), Error> {
+        let mut mul_chip = MulAddChip::construct(self.mul_gadget.clone());
+        let mut parity_check_chip = MulAddChip::construct(self.parity_check.clone());
+        let max_exp_rows = max_exp_steps * OFFSET_INCREMENT;
+        layouter.assign_region(
+            || "exponentiation circuit",
+            |mut region| {
+                region.name_column(|| "exp_table.is_step", self.exp_table.is_step);
+                region.name_column(|| "exp_table.identifier", self.exp_table.identifier);
+                region.name_column(|| "exp_table.is_last", self.exp_table.is_last);
+                region.name_column(|| "exp_table.base_limb", self.exp_table.base_limb);
+                region.name_column(|| "exp_table.exponent_lo_hi", self.exp_table.exponent_lo_hi);
+                region.name_column(
+                    || "exp_table.exponentiation_lo_hi",
+                    self.exp_table.exponentiation_lo_hi,
+                );
+                // region.name_column(|| "q_usable", self.q_usable);
+                region.name_column(|| "mul_gadget.col0", self.mul_gadget.col0);
+                region.name_column(|| "mul_gadget.col1", self.mul_gadget.col1);
+                region.name_column(|| "mul_gadget.col2", self.mul_gadget.col2);
+                region.name_column(|| "mul_gadget.col3", self.mul_gadget.col3);
+                region.name_column(|| "mul_gadget.col4", self.mul_gadget.col4);
+                // region.name_column(|| "mul_gadget.overflow", self.mul_gadget.overflow);
+                region.name_column(|| "parity_check.col0", self.parity_check.col0);
+                region.name_column(|| "parity_check.col1", self.parity_check.col1);
+                region.name_column(|| "parity_check.col2", self.parity_check.col2);
+                region.name_column(|| "parity_check.col3", self.parity_check.col3);
+                region.name_column(|| "parity_check.col4", self.parity_check.col4);
+                // region.name_column(|| "parity_check.overflow", self.parity_check.overflow);
+                debug_assert!(
+                    max_exp_rows >= self.minimum_rows,
+                    "max_exp_rows: {}, minimum_rows: {}",
+                    max_exp_rows,
+                    self.minimum_rows,
+                );
+                let mut offset = 0;
+                for exp_event in exp_events.iter() {
+                    self.assign_exp_event(
+                        &mut region,
+                        &mut offset,
+                        exp_event,
+                        &mut mul_chip,
+                        &mut parity_check_chip,
+                    )?;
+                }
+                assert!(offset <= max_exp_rows - UNUSABLE_EXP_ROWS);
```

Comment:
assertions should preferably be done before any assignment (to not assign unnecessarily)